# SMC_RNAseq 

## Experimental setup

Code for processing RNAseq data from Moushumi Das with the following _C. elegans_ strains:

- PMW366 (hs::TEV)
- PMW382 (*dpy-26*::TEVcs, hs::TEV) 
- PMW775 (*kle-2*::TEVcs, hs::TEV)
- PMW784 (*scc-1*::TEVcs, hs::TEV)
- PMW828 (*coh-1*::TEVcs, hs::TEV)
- PMW844 (*scc-1*::TEVcs, *coh-1*::TEVcs, hs::TEV)
- PMW822 (*dpy-26*::TEVcs, hs::TEV, *sdc-3*AID, P*eft-3*::TIR1)
- PMW821 (hs::TEV, *sdc-3*AID, P*eft-3*::TIR1)
- PMW823 (hs::TEV, P*eft-3*::TIR1)

All strains were heatshocked for 30 min at 34C as L1s and collected 19 hours later (L3s). The degron strains were also incubated with or without auxin as specified in the sampleTable generated by the R scripts.


## Pipeline overview and setup

The pipeline performs genomic alignment with STAR to get genome wide bigwig files, and transcriptome alignment with Salmon in order to get gene-level counts. Differential expression is evaluated with DESeq2.

The initial read mapping and counting with STAR and salmon is designed to be carried out in parallel on the slurm server. DESeq2 analysis is then carried out in R on the desktop. 

When running the pipeline for the first time (ubelix or izb cluster), the _**indexGenomeTranscripts.sh**_ needs to be run to index both the genome for STAR and the transcriptome for salmon.

## Mapping

The reads are mapped with the _**mapRNAreads.sh**_ script using STAR and Salmon.  This requires a file **fastqList.txt** (adapt the **fastqList_example.txt** file) with the following columns: 

- fileName1: full path to read1 

- fileName2: full path to read2 (NA if single end reads) 

- sampleName: name of strain used

- repeatNum:  batch and number of techincal replicate

- laneNum: number of the lane on the flow cell in which sequencing was done

- date: date when sequencing was performed

Processed files are be given unique identifiers using the sampleName,repeatNum and laneNum.

## Differential expression analysis setup

To run the DESeq2 analysis in R, the variableSettings.sh file (see **variableSettings_example.sh**) is required, which sets up how to filter the data, and what contrasts to make (including advanced contrasts).

### Custom variables

The following variables need to be changed in variableSettings.R to choose what analysis to run:

```
plotPDFs=F    # some pdfs (e.g. of correlations) can be vary large - can choose to avoid plotting pdfs and only pngs
padjVal=0.05  # general adjusted p value threshold to use for significance
lfcVal=0.5    # general log2 fold change threshold to use for significance
fileNamePrefix=paste0("p",padjVal,"_lfc",lfcVal,"/no775B3_")  # prefix to add to file name to distinguish one run from another
outPath="."         # working directory
genomeVer="WS275"   # genome version used for metadata downloaded from wormbase
genomeDir=paste0("~/Documents/MeisterLab/GenomeVer/",genomeVer)   # directory where genome data is stored
fileList<-read.table(paste0(outPath,"/fastqList_no775B3.txt"),stringsAsFactors=F,header=T)  # file with list of fastqFiles in the dataset with some metadata

rnaType="mRNA"  # perform DESeq2 analysis on reads mapped by salmon to mRNA, tnRNA, ncRNA or pseudoRNA
remakeFiles=F   # remake publicData files? unless changes are made, it is enough to do this once
combineChrAX=F  # artificially combine chrA and X from different datasets?  (Need to first run full genome and chrX only separately)
filterData=T    # filter by certain gene lists such as oscillating genes, or only chrX
filterBy=c("Cycling_Meeuse","Cycling_Latorre")    # names in filterList of gene lists to use
customNameTxt=ifelse(rnaType=="mRNA","",rnaType)  # Some other text you want to add to the filename prefix
```

These variables make it possible to run different analyses (e.g. varying the p value and LFC threshold) in the same main directory while putting the results in subdirectories. During exploratory analysis we varied both padjVal and lfcVal, and settled on the values above as ones that maximise the number of X-linked genes significantly upregulated in the dpy-26cs while minimising the number of autosomal genes upregulated.

Note: replicted 775B3 was removed from the analysis (it is absent from the fastqList_no77B3.txt file) as it was a strong outlier compared to the other thre replicates.

The final data in the paper was run filtering out oscillating genes from two datasets ("Cycling_Meeuse","Cycling_Latorre").

In addition, the scripts were run three times, once for the whole genome (using DESeq2analysis_Salmon.R script):

```
combineChrAX=F  # artificially combine chrA and X from different datasets?  (Need to first run full genome and chrX only separately)
filterData=T    # filter by certain gene lists such as oscillating genes, or only chrX
filterBy=c("Cycling_Meeuse","Cycling_Latorre")    # names in filterList of gene lists to use
```

Once for the autosomes only (using DESeq2analysis_Salmon.R script):

```
combineChrAX=F  # artificially combine chrA and X from different datasets?  (Need to first run full genome and chrX only separately)
filterData=T    # filter by certain gene lists such as oscillating genes, or only chrX
filterBy=c("Cycling_Meeuse","Cycling_Latorre", "chrX")    # names in filterList of gene lists to use
```

And finally combining the gene counts for the autosomal genes from the autosomal-only run with the gene counts for X-linked genes from the full genome run (using the combine_chrAchrX.R script):

```
combineChrAX=T  # artificially combine chrA and X from different datasets?  (Need to first run full genome and chrX only separately)
filterData=T    # filter by certain gene lists such as oscillating genes, or only chrX
filterBy=c("Cycling_Meeuse","Cycling_Latorre")    # names in filterList of gene lists to use
```

The results tables from this final run are what were used in the paper. The reason for this is the the *dpy-26*cs sample that leads to upregulation of so many X-linked genes violates the assumptions of the DESeq2 model that most genes do not change, and so to compensate for so many X-linked genes being upregulated, most autosomal genes are considered as downregulated. Something that is clearly not the case when you analyse the autosomes on their own.

### Advanced contrasts

We initially performed the analysis separately on different batches: the klesin cleavage sample and the degron samples (which included *dpy-26*cs with and without the *sdc-3*AID), but to avoid having two different sets of numbers for significant genes for the same *dpy-26*cs sample due to slight variations of the DESeq2 model for each batch, we performed the final analysis for all the samples together (the results were similar). As our range of RNAseq conditions did not completely cover all possible combinations of conditions (kleisin cleavage and auxin treatment) we created a variable composed of the combinations seen in our data, by combining the status for which kleisin with a cut site was present, presence of TIR protease, presence of *sdc-3*AID, and auxin treatment into a single "SMC" factor variable with 11 levels:

| strain    | kleisin    | TIR1 | sdc3    | auxin | SMC                      |
|-----------|------------|------|---------|-------|--------------------------|
| PMW366    | wt         | wt   | wt      | 0mM   | wt.wt.wt.0mM             |
| PMW366    | wt         | wt   | wt      | 1mM   | wt.wt.wt.1mM             |
| PMW382    | dpy26cs    | wt   | wt      | 0mM   | dpy26cs.wt.wt.0mM        |
| PMW775    | kle2cs     | wt   | wt      | 0mM   | kle2cs.wt.wt.0mM         |
| PMW784    | scc1cs     | wt   | wt      | 0mM   | scc1cs.wt.wt.0mM         |
| PMW821    | dpy26cs    | TIR1 | sdc3deg | 1mM   | dpy26cs.TIR1.sdc3deg.1mM |
| PMW823    | wt         | TIR1 | wt      | 1mM   | wt.TIR1.wt.1mM           |
| PMW822    | wt         | TIR1 | sdc3deg | 1mM   | wt.TIR1.sdc3deg.1mM      |
| PMW822    | wt         | TIR1 | sdc3deg | 0mM   | wt.TIR1.sdc3deg.0mM      |
| PMW828    | coh1cs     | wt   | wt      | 0mM   | coh1cs.wt.wt.0mM         |
| PMW844    | scc1coh1cs | wt   | wt      | 0mM   | scc1coh1cs.wt.wt.0mM     |


The basic DESeq2 model was built with the PWM366 sample as the reference "control" sample. However for the degron samples this was not always the most appropriate control (e.g. for the PMW821 strain (hs::TEV, *sdc-3*AID, P*eft-3*::TIR1) we wanted to look at the +- auxin conditions. Advanced contrasts also helped us examine marginal effects such as the effect of *sdc-3*AID in the presence of *dpy-26*cs. These contrasts are optionally created towards the end of the variableSetting.R script by subtracting coefficients of models built with the different levels of the SMC variable above. The contrast name scheme uses the same positional variable scheme as the SMC variable above, but an X is placed in the position of the conditions that vary, and the specific subtraction is written after an \_ at the end of the contrast name. Although not all were used in the paper, the following contrasts were examined:

| contrastName                               | shortName       | description                                                                   | In paper |
|--------------------------------------------|-----------------|-------------------------------------------------------------------------------|-------|
| wt.wt.wt.X\_1mM\_vs\_0mM                   | aux             | Effect of 1mM auxin in wt background                                          | No    |
| wt.TIR1.sdc3deg.X\_1mM\_vs\_0mM            | aux_sdc3BG      | **Effect of SDC-3 degradation +-auxin**                                           | **Yes**   |
| wt.X.wt.1mM\_TIR1\_vs\_wt                  | TIR1            | Effect of TIR1 in the presence of auxin                                       | No    |
| wt.X.wt.X\_TIR11mM\_vs\_wt0mM              | TIR1aux         | Effect of TIR1 and auxin                                                      | No    |
| wt.TIR1.X.1mM\_sdc3deg\_vs\_wt             | sdc3            | Effect of +- *sdc-3*AID transgene (with auxin always preseent)                | No    |
| X.wt.wt.0mM\_dpy26cs\_vs\_wt               | **dpy26**       | **Effect of DPY-26 cleavage**                                                 | **Yes**   |
| X.TIR1.sdc3deg.1mM\_dpy26cs\_vs\_wt        | dpy26_sdc3BG    | Effect of DPY-26 cleavage when SDC-3 is degraded                              | No    |
| dpy26cs.X.X.X\_TIR1sdc3deg1mM\_vs\_wtwt0mM | sdc3_dpy26BG    | Effect of *sdc-3*AID+TIR1+auxin when DPY-26 is cleaved                        | No    |
| X.TIR1.X.1mM\_dpy26cssdc3deg\_vs\_wtwt     | dpy26sdc3       | Effect of +- *dpy-26*cs and *sdc-3*AID transgenes in presence of TIR1 & auxin | No    |
| X.TIR1.sdc3deg.X\_dpy26csaux\_vs\_wt0mM    | **aux_dpy26sdc3**| **Effect of DPY-26 cleavage and SDC-3 degradation (+- auxin)**                | **Yes**   |
| **X.wt.wt.0mM\_kle2cs\_vs\_wt                | kle2**            | **Effect of KLE-2 cleavage**                                                  | **Yes**   |
| **X.wt.wt.0mM\_scc16cs\_vs\_wt               | scc1**            | **Effect of SCC-1 cleavage**                                                  | **Yes**   |
| **X.wt.wt.0mM\_coh1cs\_vs\_wt                | coh1**            | **Effect of COH-1 cleavage**                                                  | **Yes**   |
| **X.wt.wt.0mM\_scc1coh1cs\_vs\_wt            | scc1coh1**        | **Effect of SCC-1 and COH-1 cleavage**                                        | **Yes**   |
| X.wt.wt.0mM\_scc1coh1cs-coh1               | scc1coh1cs-coh1 | Effect of SCC-1 cleavage in the presence of cleaved COH-1                     | No    |
| X.wt.wt.0mM\_scc1coh1cs-scc1               | scc1coh1cs-scc1 | Effect of COH-1 cleavage in the presence of cleaved SCC-1                     | No    |



## Differential expression analysis
The output is then analysed in R with DESeq2 using the following script:

**_DESeqAnalysis_SALMON.R_** 

This produces the following plots:

- basic QC plots: bar plots and density plots of the raw counts,sample-sample clustering heatmap, heatmap of top 500 expressed genes with and without sample clustering, PCA plots coloured by main variables (_./plots/salmon_sampleQC.pdf_)

- boxplots of lfc of genes by chr or chr type (_./plots/salmon_xxxx_boxplots_expnByChr.pdf_, _./plots/salmon_xxxx_boxplots_expnByChrType.pdf_)

- MAplots (_./plots/salmon_xxxx_MAplots_results.pdf_)

- Volcano plots (_./plots/salmon_xxxxx_volcanoPlot_xxxxx.pdf_)

- heirarchical clustering of samples by most changed genes (_./plots/salmon_xxxxxx_hclust_mostChanged.pdf_)

- barplots of expression of a few individual genes, most changed in each sample (_./plots/salmon_xxxx_topGenes_normCounts.pdf_)

The script also produces tables of results with the LFC and pvalues for contrasts specified in variableSettings.R

## Further exploratory analysis
### Comparison of the different data sets

The significant genes from each data set are compared by:
_**compareDatasets.R**_

This script produces the following plots:

- barplots to campare counts of genes per chromosome (_./plots/bar_countsPerChr_xxxxxx.pdf_)

- venn diagrams to compare overlap between datasets (_./plots/venn_xxxxxx.pdf_)

- correlation of the log2 fold change of genes shared between data sets (_./plots/cor_xxxxx.png_)

### Comparison of HiC features

The association of gene expression with different HiC features are compared by:
_**compareHICfreatures.R**_


### Comparison of Tissue and GO term enrichment

_**compareTissueAndGO.R**_


### Publically available datasets

These datasets were retrieved manually:

**DCgenes_published.xlsx** 

**Jans2009_DC_suplTable4.txt**

**Jans2009_notDC_suplTable5.txt** 

**Kramer et al. (2015)** was retrieved automatcally.

All were processed with:

_**processPublished.R**_
